components/sup/src/manager/commands.rs

```diff
@@ -185,81 +184,39 @@ pub fn service_load(mgr: &ManagerState,
-    let bldr_url = opts.bldr_url
-                       .clone()
-                       .unwrap_or_else(|| protocol::DEFAULT_BLDR_URL.to_string());
-    let bldr_channel = opts.bldr_channel
-                           .clone()
-                           .map(ChannelIdent::from)
-                           .unwrap_or_default();
-    let force = opts.force.unwrap_or(false);
     let source = InstallSource::Ident(ident.clone(), PackageTarget::active_target());
-    match mgr.cfg.spec_for_ident(source.as_ref()) {
-        None => {
-            let mut spec = ServiceSpec::default();
-            opts.into_spec(&mut spec);
-
-            // We don't have any record of this thing; let's set it up!
-            //
-            // If a package exists on disk that satisfies the
-            // desired package identifier, it will be used;
-            // otherwise, we'll install the latest suitable
-            // version from the specified Builder channel.
-            let package = util::pkg::satisfy_or_install(req, &source, &bldr_url, &bldr_channel)?;
-            if let Err(e) = spec.validate(&package) {
-                return Err(net::err(ErrCode::InvalidPayload,
-                                    format!("Failed to validate '{}' due to '{}'",
-                                            ident, e)));
-            }
-
-            mgr.cfg.save_spec_for(&spec)?;
-            req.info(format!("The {} service was successfully loaded", spec.ident))?;
+    let mut spec = if let Some(spec) = mgr.cfg.spec_for_ident(source.as_ref()) {
+        let force = opts.force.unwrap_or(false);
+        // We've seen this service  before. Thus `load`
+        // basically acts as a way to edit spec files on the
+        // command line. As a result, we a) check that you
+        // *really* meant to change an existing spec, and b) DO
+        // NOT download a potentially new version of the package
+        // in question
+        if !force {
+            return Err(net::err(ErrCode::Conflict,
+                                format!("Service already loaded, unload '{}' \
+                                         and try again",
+                                        ident)));
         }
-        Some(mut spec) => {
-            // We've seen this service  before. Thus `load`
-            // basically acts as a way to edit spec files on the
-            // command line. As a result, we a) check that you
-            // *really* meant to change an existing spec, and b) DO
-            // NOT download a potentially new version of the package
-            // in question
-
-            if !force {
-                return Err(net::err(ErrCode::Conflict,
-                                    format!("Service already loaded, unload '{}' \
-                                             and try again",
-                                            ident)));
-            }
+        spec
+    } else {
+        ServiceSpec::default()
+    };
```

Comment:
One way to consolidate some of this might be:

```rust
let mut spec = mgr.cfg.spec_for_ident(source.as_ref()).unwrap_or_default();
if spec.force.unwrap_or(false) { // or spec.force.unwrap_or_default()
  return Err(...);
}
```

Reply:
:thinking: I am not sure this will work? We only want to check for `force` if the spec already existed. This would always check for it and require the user to always specify `--force`. Are you saying we should store `force` in the spec? Currently, it is set in `opts`.

Reply:
I agree that this won't really work. You'd have to pass the `force` opt into `mgr.cfg.spec_for_ident` or something to get the appropriate behavior, but I'm not sure that's what we want.

Alternatively, if there were a function that just checked for the existence of a spec, we might be able to make things nicer, but overall this logic is so convoluted that I don't think can be easily simplified. Before we even _try_, I'd want to get the existing behavior documented via tests so it can be more safely refactored.

Reply:
ah, correct... :grimacing: